### DISCUSSION TITLE
FR: Allow multiple instances of a configured field in a single field layout

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#8497] [Ideas] [answered]
This would likely be a big change, but thought I'd throw it out there, perhaps to be considered with https://github.com/craftcms/cms/issues/1524

------

Field configuration can be complex, especially with a heavily configured Matrix field.
It would be ideal if one could create a field and use multiple instances of it in a single field layout.

### Use-case:
A fairly common practice for module page building seems to be including a mega "Content Matrix" field, with many block types for all types of content. This has a breaking point when you need two independent sets of these blocks (e.g. _Content Matrix Main_, _Content Matrix Sidebar_).

Duplicating a second field is tough, and keeping them in-sync over time is even harder.

This could potentially also be a solution for issues like https://github.com/craftcms/cms/issues/806, as each field instance could have it's own title and instructions.

I toyed around with the idea of creating a plugin with a "Proxy FieldType", where the only field setting was a field selection of the field you wanted to mirror. Works great for things stored in `content`, but obviously gets complicated with Matrix and others, which is really what you'd want it for. Let me know if you think this might be do-able and/or more appropriate via plugin, or if it is something you'd consider for core/v4/fantasy land.

**Comment by khalwat** [comment]
What I've done for this in the past is creating a Field Group that encompasses everything that should be added to a new section. There are a couple of issues with this, however.

1. If you change the Field Group, it doesn't change the fields actually added to the section(s); it's really just organizational.
2. You can't then use these Field Groups in a Matrix block

I think this is one of the reasons why Neo is popular, it lets you use field layouts with block content... and re-use them.

Part of the flexibility of Craft is that you can re-use fields, and you can add/remove fields from any Section layout or matrix block.

Still, it'd be pretty interesting to have a "Field Container" or such which would be a collection of arbitrary fields, but in the Field Layout editor, it's treated as one chunk that's always moved together.

If you want to change what fields are in it, you change those in the Field Container settings... and any changes made there would globally affect any instances of that Field Container.

Really, we're just looking for a way to group fields together in more than just an organizational manner (which is still useful in its own right).

**Comment by brandonkelly** [comment]
Agree it would be nice.

**Comment by foxtrotcz** [comment]
I just want to say, I am really looking forward for this feature.
A lot of time I am using simply configured fields like plain text or table and I would like to use more instances of that fields in field layout.
For example. If you have plain text field Email and you want to use it multiple times, it makes more sense to use multiple instances of same field (inside field layout) than configuring multiple fields like Email1, Email2, Email3...

Thanks, this would be great feature.

**Comment by turnstylerj** [comment]
Bumping this discussion as the newer capability of being able to apply custom labels to reused fields in the layout designer made me think of how nice it'd be to be able to re-use fields giving them custom handles or something, but had no idea how it could be implemented. I think OP's idea of a proxy field would be a great solution.

To throw in another use-case: we typically run into this issue when mixing common matrix blocks with entry-level fields, particularly with singles/unique pages that also contain shared matrix blocks. If the unique fields are only above or below the matrix blocks, we can set up the layout like so:

![entry-fields-01-sm](https://user-images.githubusercontent.com/42223501/170391600-cdc6e7d5-74e7-4036-b949-8f8a486fc104.jpg)

However, it gets dicey when the unique entry-level fields are in-between matrix blocks, like so:

![entry-fields-02-sm](https://user-images.githubusercontent.com/42223501/170391638-1739d42b-5f0c-41b3-914b-92b7ad887ad1.jpg)

In situations like this, it would really clean up the experience if we were able to just somehow re-use the same matrix field in the field layout designer above and below the unique entry fields, rather than clutter up the shared matrix with blocks only used in a single location.

**Comment by alexanderloewe** [comment]
I would like to see that.

**Reply by TheFunkyMonk** [reply]
Forgot about this issue, but I’m pretty sure I remember hearing it’s planned for Craft 5! @brandonkelly is that right?

**Reply by brandonkelly** [reply]
Yep! Added for Craft 5

**Comment by brandonkelly** [comment, marked as answer]
Craft 5 is out with the ability to create multiple instances of individual fields within a field layout. 🚀

**Comment by sandrodunkel** [comment]
Is there any specific reason why this does not work with asset or matrix fields?

**Reply by brandonkelly** [reply]
Matrix and relation fields each have their own ways of storing content (as elements, and as rows in the `relations` table, respectively). So they don’t get multi-instance support for free like other fields that just let Craft store their values in the `content` column.

It might be possible for us to add multi-instance support to them, but it’s going to be a fair bit of work, so it didn’t make sense to try to tackle it in the initial multi-instance fields effort.

**Reply by dolphiq-allard** [reply]
Fair enough, thanks @brandonkelly for the clarification. Maybe in a future release.

**Reply by brandonkelly** [reply]
Relational fields (Assets, Entries, etc.) are gaining multi-instance support in Craft 5.3 (#15400) 🎉

**Reply by ryanleichty** [reply]
I noticed that some matrix fields allow multi-instance (with say two instances, but not more), while most others don't. Will matrix fields eventually gain full multi-instance support?

**Reply by brandonkelly** [reply]
@ryanleichty Matrix fields don’t allow two instances, just one. Not sure how you’d end up with two instances of the same Matrix field, but that’s going to lead to bugs. We’re not currently planning on making it possible.

**Comment by brandonkelly** [comment]
Craft 5.3.0-beta.1 is out with support for multi-instance relational fields 🎉 (#15400)

We’re targeting August 6 for the GA release. If you’d like to test ahead of time, change your `craftcms/cms` requirement in composer.json to:

```json
"craftcms/cms": "^5.3.0-beta.1",
```

and then run `composer update`.

**Reply by nstcactus** [reply]
I'm running Craft 5.10.8.1 and I cannot add more than one instance of a matrix field to a entry type's field layout. 

What did I do wrong?

**Reply by mmikkel** [reply]
@nstcactus You're not doing anything wrong; unlike most other field types Matrix fields do not support multiple instances at this time.

The workaround for now, is to create multiple Matrix fields, using the same entry types.

**Reply by nstcactus** [reply]
Thanks for your answer.

Then I didn't understand Brandon's message:

> Craft 5.3.0-beta.1 is out with support for multi-instance relational fields 🎉 (#15400)

I thought matrix fields were relational fields

**Reply by brandonkelly** [reply]
No, Matrix fields allow you to _create_ nested entries.

Relational fields (Entries, Assets, Users, etc.) allow you to define _relationships_ with other elements. But if you delete the relation, the target element doesn’t get deleted as well.

**Comment by BenHMatrix** [comment]
Is the new Content Block restricted to a single instance in the same way as Matrix fields?
I was really hoping to have a "Colours" Content Block field (with sub fields for Heading Colour, Text Colour and Background Colour) and re-use that across Global sets etc. But I can only add the Colours Content Block field once. 

If not this purpose (combining fields into reusable groups), what is the Content Block field for?

**Reply by brandonkelly** [reply]
@BenHMatrix Correct, it doesn’t support multiple instances, for the same reason as Matrix. I did attempt to make it multi-instance, but it got a little too hacky/over-engineered so I bailed.

See the related issues referenced in #17424 for plenty of examples of use cases for Content Block fields.

**Reply by BenHMatrix** [reply]
Thanks!
